raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1487,39 +1510,57 @@ public void poll() throws IOException {
     }
 
     private void handlePendingAppends(long currentTimeMs) {
-        UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
-        if (unwrittenAppend == null || unwrittenAppend.isCancelled())
-            return;
-
         if (quorum.isLeader()) {
-            if (unwrittenAppend.isTimedOut(currentTimeMs)) {
-                unwrittenAppend.fail(new TimeoutException());
-            } else {
-                LeaderState leaderState = quorum.leaderStateOrThrow();
-                int epoch = quorum.epoch();
-                LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
-                if (info != null) {
-                    OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
-                    unwrittenAppend.complete(offsetAndEpoch);
-                    uncommittedAppends.put(offsetAndEpoch, new AppendBatchAndTime(info.lastOffset - info.firstOffset + 1, currentTimeMs));
+            for (UnwrittenAppend unwrittenAppend: unwrittenAppends) {
```

Comment:
I found that with a blocking array of size 10, it is possible to fill in the size and cause unnecessary failures. So I've decided to 1) use a linked queue with no capacity limit, and 2) try to drain all of them in each call instead of only one at a time. LMK what do you think.

Reply:
I agree it seems like a better policy than only draining one. I wonder if there are any concerns from a liveness perspective. Could this lead to thread starvation if the appends are coming fast enough that we can't drain them all. I wonder if it would make sense to set a maximum count that can be drained on each poll?

nit: space before colon